src/TokenAirdrop.hs

```diff
@@ -43,16 +49,21 @@ tokenAirdrop config = do
       addrs = address <$> beneficiaries
       pubKeyAddressMap :: Map PubKeyHash PubKeyAddress
       pubKeyAddressMap = fromList $ zip (pkaPubKeyHash <$> addrs) addrs
-      indexedTxs :: [(Constraints.TxConstraints Void Void, [Beneficiary], Int)]
+      indexedTxs :: [IndexedTransaction]
       indexedTxs = zipWith combine2To3 txPairs [1 :: Int ..]
 
-  if config.live
-    then do
-      confirm <- confirmTxSubmission
-      if confirm
-        then processTransactions indexedTxs pubKeyAddressMap
-        else pure $ Left "Operation stopped by user"
-    else processTransactions indexedTxs pubKeyAddressMap
+  runExceptT $ do
+    when config.live $ do
+      confirmed <- lift confirmTxSubmission
+      unless confirmed $ throwError "Operation stopped by user"
+
+    ExceptT $
+      processTransactions indexedTxs pubKeyAddressMap
+        >>= \case
+          Right _ -> pure $ Right ()
+          Left (msg, txs) -> do
+            logBeneficiares $ fmap (\(_, b, _) -> b) txs
+            pure $ Left msg
```

Comment:
A bit ugly, but mtl helps a lot with combining the cases. In my token-sum-check branch, it's also going to be useful as there will be yet another check we're performing before actually processing the transactions.